<a href="https://colab.research.google.com/github/nedokormysh/Stepik_LLM/blob/week_2_prompting/LLM_M2_1_Prompt_Engeneering_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <h1><center> 🤹‍♀️ Prompt Engeneering - выжми из модели максимум </center></h1>


<img src='https://cdn.practical-tips.com/wp-content/uploads/2023/07/7s.jpeg' align="right" width="528" height="528" >

## Оглавление ноутбука:

 * <a href="#c1"> 🚀Введение  </a>
   * <a href="#c2"> ❄️ Структура промпта </a>
* <a href="#look1"> 🌡 Generation Temperature - настраиваем "творческость" модели!  </a>
* <a href="#check1"> 👯‍♀️ Добавление нескольких контекстов  </a>
* <a href="#check2"> 🧐 Как посчитать расходуемые токены и деньги 💸? </a>
* <a href="#check3"> 🤹‍♀️ Техники Prompt Engeneering и кто такой Prompt Engeneer 👨‍🔧?</a>
* <a href="#6">🧸 Выводы и заключения</a>

<div class="alert alert-info">

В этом ноутбуке попытаемся разобраться что такое `Prompt Engeneering` (дизайн промпта) и почему так важно в нём разобраться в первую очередь, начиная изучение сферы LLM моделей.<br>
*Начнём с определений:*

***Промпт (prompt)*** - это запрос, который мы подаем большой языковой модели, чтобы получить нужную информацию или выполнить определенную задачу. Он определяет, как именно LLM будет интерпретировать нашу просьбу и формировать ответ. По сути, хорошо сформулированный промпт выступает своего рода ключом, открывающим доступ к возможностям языковой модели.

***Prompt Engeneering (дизайн промпта)*** — это практика разработки и уточнения промптов (вопросов или инструкций) для получения конкретных ответов от моделей ИИ (не только языковых, но и от других, например, генерации картинок). Думайте об этом как об интерфейсе между намерениями человека и результатами работы машины. В обширной сфере искусственного интеллекта, где модели обучаются на огромных наборах данных, правильный промпт может стать решающим фактором между пониманием моделью вашего запроса или его неверной интерпретацией. Например, если вы когда-либо взаимодействовали с голосовыми помощниками, такими как `Siri` или `Алиса`, вы занимались базовой формой `Prompt Engeneering'а`.

На примере ниже, вы можете увидеть, как незначительные изменения в тексте промпта, могут значительно изменить результат генерации. <br>
Запросы к генеративной нейросети [Midjourney](https://docs.midjourney.com/docs/multi-prompts).

<img src='https://github.com/a-milenkin/LLM_practical_course/blob/main/images/cheesecake.png?raw=1' align="center">

## ❄️ Структура промпта

<div class="alert alert-info">

Компоненты, из которых может быть составлен промпт:

* **Инструкции (Instructions)** - Этот компонент обеспечивает модели ясные указания о том, что ожидается от нее в ответе. Инструкция определяет, как модель должна интерпретировать введенные данные и какие параметры ответа следует учесть. Например, инструкция может содержать информацию о стиле ответа, длине текста или других ограничениях.
* **Внешний контекст (External information or context)** - Этот компонент предоставляет модели дополнительный контекст или информацию, которую она может использовать при формировании ответа. Это может включать в себя факты, данные, или ссылки на внешние источники информации, которые могут быть полезными для модели. Например, если вы общались с `ChatGPT`, то вся история вашего диалога сохранялась и использовалась в качестве контекста при каждом новом сообщении, чтобы модель понимала о чём вы с ней разговаривали и не перескакивала с темы на тему.
* **Ввод пользователя или запрос (User input or query)** - Это входные данные, которые пользователь предоставляет модели. Это может быть вопрос, просьба, или какой-либо запрос, который пользователь хочет, чтобы модель обработала.
* **Выходной индикатор (Output indicator)** -  Этот компонент указывает, как модель должна сформировать свой ответ. Это может быть, например, просьба о предоставлении ответа в определенной форме или формате, такой как краткое резюме, расширенное объяснение, код, или что-либо еще.

Не всегда нужно указывать все компоненты для хорошего промпта, но помните, что корректное использование и балансирование этих компонентов помогает точно настраивать модель и получать желаемые результаты.

In [ ]:
import os
from getpass import getpass
import warnings
warnings.filterwarnings('ignore')

In [ ]:
!pip install langchain openai tiktoken -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 19.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.0/77.0 kB 7.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 35.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.5/44.5 kB 4.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 4.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.


In [ ]:
# Для работы в колабе загрузите наш скрипт для использования ChatGPT на сервере курса!
!wget https://raw.githubusercontent.com/a-milenkin/LLM_practical_course/main/notebooks/utils.py

--2023-11-04 17:11:23--  https://raw.githubusercontent.com/a-milenkin/LLM_practical_course/main/notebooks/utils.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9404 (9.2K) [text/plain]
Saving to: ‘utils.py’

utils.py            100%[===================>]   9.18K  --.-KB/s    in 0s      

2023-11-04 17:11:23 (33.5 MB/s) - ‘utils.py’ saved [9404/9404]



In [ ]:
# # Если используете ключ от OpenAI, запустите эту ячейку
# from langchain.chat_models import ChatOpenAI

# # os.environ['OPENAI_API_KEY'] = "Введите ваш OpenAI API ключ"
# os.environ['OPENAI_API_KEY'] = getpass(prompt='Введите ваш OpenAI API ключ')

# # инициализируем языковую модель
# llm = ChatOpenAI(temperature=0.0)

In [ ]:
# Если используете ключ из курса, запустите эту ячейку
from utils import ChatOpenAI

#course_api_key= "Введите ваш OpenAI API ключ"
course_api_key = getpass(prompt='Введите API ключ')


# инициализируем языковую модель
llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key)

Введите API ключ··········


Напишем промпт

In [ ]:
prompt = """Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь 'Я не знаю'.

Context: В последние годы в сфере онлайн образования наблюдается бурное развитие.
Открывается большое количество платформ для хостинга курсов.
Одни из самых крупных платформ в мире, это Coursera и Udemi.
В России лидером является Stepik.

Question: На каких онлайн платформах можно размещать курсы?

Answer: """

<div class="alert alert-info">

В этом примере мы имеем:
* Инструкцию
* Контекст
* Вопрос (запрос пользователя)
* Выходной индикатор ("Answer: ")

In [ ]:
print(llm.predict(prompt))

На Coursera, Udemi и Stepik.


В качестве альтернативы, если у нас нет нужной информации в `контексте`, модель должна ответить `"Я не знаю"`, давайте попробуем.

In [ ]:
prompt = """Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь 'Я не знаю'.

Context: В Интернете много сайтов.

Question: На каких онлайн платформах можно размещать курсы?

Answer: """


In [ ]:
print(llm.predict(prompt))

Я не знаю.


<div class="alert alert-info">

Отлично, модель понимает наши инструкции. В большинстве реальных случаев мы не будем предоставлять внешнюю информацию/контекст модели вручную. Это будет автоматический процесс, использующий что-то вроде долговременной памяти для извлечения соответствующей информации из внешнего источника (об этом подробнее в 4-м модуле курса).


## <center> 🌡 Generation Temperature - настраиваем "творческость" модели! </center>

Параметр `temperature`, используемый в моделях генерации, говорит нам, насколько «случайной» может быть модель. Он представляет собой вероятность того, что модель выберет слово, которое не является первым выбором модели. Это работает, потому что модель фактически назначает прогноз вероятности для всех токенов в своем словаре на каждом «шаге» модели (каждом новом слове или подслове).

<div class="alert alert-info">
Вкратце, увеличение температуры может привести к большей случайности, что способствует более разнообразным или творческим результатам. Вы фактически увеличиваете веса других возможных токенов. В плане применения, для задач, связанных с ответами на вопросы на основе фактов, рекомендуется использовать более низкое значение температуры, чтобы стимулировать более точные и краткие ответы. Для генерации стихов или других творческих задач может быть полезно увеличить значение температуры.

Давай попробуем:

In [ ]:
prompt = """Ниже представлен диалог с весёлым чатботом.
Ответы чатбота остроумные и забавные.

Chatbot: Привет! Я чатбот.
User: Привет, как дела?
Chatbot: """

llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key)
print(llm.predict(prompt))

Лучше всех! Я ведь чатбот, у меня всегда все отлично. А у тебя как дела?


In [ ]:
llm = ChatOpenAI(temperature=1.0, course_api_key=course_api_key)
print(llm.predict(prompt))

Прекрасно! Я всегда в отличном настроении, потому что я не имею эмоций.


<div class="alert alert-info">
Второй ответ более креативен и демонстрирует разницу, которую мы можем ожидать между низкотемпературными и высокотемпературными генерациями.

## <center> 👯‍♀️ Добавление нескольких контекстов </center>

<div class="alert alert-info">

В некоторых юзкейсах, таких как вопрос-ответ, мы можем использовать внешний источник информации, чтобы повысить надежность или фактологичность ответов модели. Мы называем эту информацию `source knowledge`, то есть любыми знаниями, введенными в модель через промпт.

Создадим список «фиктивной» внешней информации, воспользуемся строчками из документации `OpenAI`.

In [ ]:
contexts = [
    (
        "Large Language Models (LLMs) are the latest models used in NLP. " +
        "Their superior performance over smaller models has made them incredibly " +
        "useful for developers building NLP enabled applications. These models " +
        "can be accessed via Hugging Face's `transformers` library, via OpenAI " +
        "using the `openai` library, and via Cohere using the `cohere` library."
    ),
    (
        "To use OpenAI's GPT-3 model for completion (generation) tasks, you " +
        "first need to get an API key from " +
        "'https://beta.openai.com/account/api-keys'."
    ),
    (
        "OpenAI's API is accessible via Python using the `openai` library. " +
        "After installing the library with pip you can use it as follows: \n" +
        "```import openai\nopenai.api_key = 'YOUR_API_KEY'\nprompt = \n" +
        "'<YOUR PROMPT>'\nres = openai.Completion.create(engine='text-davinci" +
        "-003', prompt=prompt, max_tokens=100)\nprint(res)"
    )
]

<div class="alert alert-info">

Нам нужно ввести внешнюю информацию в наш промпт между первоначальными инструкциями и пользовательским вводом. Для моделей `OpenAI` рекомендуется отделять контексты от остальной части подсказки с помощью `###` или `"""`, а каждый независимый контекст можно отделить несколькими символами новой строки `\n` и ##, например:

In [ ]:
# Воспользуемся методом join для объединения контекстов в 1 строку с разделителем
context_str = '\n\n##\n\n'.join(contexts)

# Напишем промпт
prompt = f"""Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь 'Я не знаю'.

###

Contexts:
{context_str}

###

Question: Дай мне 2 примера как использовать GPT-3 на Python

Answer: """

In [ ]:
llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key)
print(llm.predict(prompt))

1. Установить библиотеку `openai` с помощью pip:
   ```bash
   pip install openai
   ```

   Затем подключиться к API с помощью вашего API-ключа:
   ```python
   import openai
   openai.api_key = 'YOUR_API_KEY'
   ```

   Создать запрос на завершение (генерацию) текста с помощью GPT-3:
   ```python
   prompt = '<YOUR PROMPT>'
   res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)
   print(res)
   ```

2. Получить API-ключ от OpenAI на сайте 'https://beta.openai.com/account/api-keys'. Установить библиотеку `openai` с помощью pip:
   ```bash
   pip install openai
   ```

   Подключиться к API с помощью вашего API-ключа:
   ```python
   import openai
   openai.api_key = 'YOUR_API_KEY'
   ```

   Создать запрос на завершение (генерацию) текста с помощью GPT-3:
   ```python
   prompt = '<YOUR PROMPT>'
   res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)
   print(res)
   ```


<div class="alert alert-info">

Неплохо, но действительно ли эти контексты помогают? Возможно, модель способна ответить на эти вопросы без дополнительной информации (`source knowledge`), поскольку может полагаться исключительно на информацию, хранящуюся во внутренних параметрах модели (`parametric knowledge`). Давайте еще раз спросим без внешней информации.

In [ ]:
prompt = f"""Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, ответь 'Я не знаю'.

Question: Дай мне 2 примера как использовать GPT-3 на Python

Answer: """

llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key)
print(llm.predict(prompt))

1. Пример использования GPT-3 на Python для генерации текста:
```python
import openai

openai.api_key = 'YOUR_API_KEY'

prompt = "Once upon a time"
response = openai.Completion.create(
  engine="davinci-codex",
  prompt=prompt,
  max_tokens=100,
  n=1,
  stop=None
)

generated_text = response.choices[0].text.strip()
print(generated_text)
```
Этот пример демонстрирует, как использовать GPT-3 для генерации продолжения текста на основе заданного начала ("Once upon a time").

2. Пример использования GPT-3 на Python для ответа на вопросы:
```python
import openai

openai.api_key = 'YOUR_API_KEY'

context = "GPT-3 is a state-of-the-art language model developed by OpenAI."
question = "What is GPT-3?"

response = openai.Answer.create(
  search_model="davinci",
  model="davinci-codex",
  question=question,
  documents=[context],
  examples_context=context,
  max_tokens=100
)

answer = response['choices'][0]['answer']
print(answer)
```
В этом примере GPT-3 используется для отвечания на вопросы на

<div class="alert alert-info">

Это не совсем то, что мы просили, и определенно не очень конкретное. Таким образом, очевидно, что добавление некоторых исходных знаний к нашим подсказкам может привести к гораздо лучшим результатам. В следующих уроках мы научимся, как проще добалять контекст к промптам и даже использовать внешние базы знаний.

# <center> 🧐 Как посчитать расходуемые токены и деньги 💸?

<div class="alert alert-info">

Учитывая, что нам может потребоваться ввести внешнюю информацию в наши промпты, они могут стать довольно большими. Максимальное контекстное окно LLM относится к токенам как в промпте, так и в тексте ответа. Для `text-davinci-003` это 4097 токенов. Однако измерение общего количества входных токенов является более сложным.
Поскольку токены не сопоставляются непосредственно со словами, мы можем измерить количество токенов в тексте только путем фактической токенизации текста. Модели GPT используют токенизатор `TikToken` от `OpenAI`. Мы можем установить библиотеку через `Pip`:

In [ ]:
#!pip install -qU tiktoken

Используя токенайзер, можем посчитать количество токенов в предыдущем промпте:

In [ ]:
import tiktoken

prompt = f"""Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь 'Я не знаю'.

###

Contexts:
{context_str}

###

Question: Дай мне 2 примера как использовать GPT-3 на Python

Answer: """

encoder_name = 'p50k_base'
tokenizer = tiktoken.get_encoding(encoder_name)

len(tokenizer.encode(prompt))

440

<div class="alert alert-info">

Это мы получили количество токенов только в промпте, количество токенов для ответа модели регулируется праметром `max_tokens` (по умолчанию 256), указывается максимально возможная длина, т.е. не обязательно, что модель всегда будет расходовать весь доступный лимит токенов, но точно не будет превышать.

In [ ]:
llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key, max_tokens=512)
print(llm.predict(prompt))

1. Установите библиотеку `openai` с помощью pip и импортируйте ее в свой проект на Python. Затем получите API-ключ от OpenAI, используя ссылку 'https://beta.openai.com/account/api-keys'. После получения ключа, установите его в переменную `openai.api_key`.

Пример кода:
```python
import openai

openai.api_key = 'YOUR_API_KEY'

prompt = 'Ваше начальное предложение'

res = openai.Completion.create(engine='text-davinci-003', prompt=prompt, max_tokens=100)

print(res)
```

2. Если вы хотите использовать модель GPT-3 для задачи генерации текста, вы можете использовать метод `openai.Completion.create()` и передать ему параметры, такие как `engine`, `prompt` и `max_tokens`. Здесь `engine` указывает на модель GPT-3, `prompt` - это ваше начальное предложение, а `max_tokens` - это максимальное количество слов в сгенерированном тексте.

Пример кода:
```python
import openai

openai.api_key = 'YOUR_API_KEY'

prompt = 'Ваше начальное предложение'

res = openai.Completion.create(engine='text-davinci-0

In [ ]:
len(tokenizer.encode(llm.predict(prompt)))

1083

<div class="alert alert-info">

Получаем 440 + 504 ~ 1000 токенов на 1 вопрос-ответ, при стоимости 1000 токенов `0.002$`. Т.е. 10 пользователей за 100 запросов сожгут уже `2$`. Причем окно контекста используется только на 1\3. <br>
Начинаем продумывать экономику своего сервиса 🤑

# <center> 🤹‍♀️ Техники Prompt Engeneering и кто такой Prompt Engeneer 👨‍🔧?

#### **Базовые техники** (советы, которые обычный пользователь может использовать, чтобы улучшить свои промпты):
*  **Ролевая игра** Заставив модель действовать как конкретная сущность, например, историк или ученый, вы можете получить индивидуальные ответы. Например, фраза «Как диетолог, оцените следующий план диеты» может привести к ответу, основанному на науке о питании.
* **Итеративное уточнение** Начните с общей фразы запроса и постепенно уточняйте ее на основе ответов модели. Этот итеративный процесс помогает довести промпт до совершенства.
* **Цикл обратной связи** Используйте выходные данные модели для информирования и корректировки последующих промптов. Такое динамическое взаимодействие гарантирует, что ответы модели с течением времени будут более точно соответствовать ожиданиям пользователей.

#### **Продвинутые техники** (более сложные стратегии, требующие более глубокого понимания поведения модели. Будем рассматривать дальше в курсе):
*  **Zero-shot prompting** Этот метод предполагает предоставление модели задачи, которую она не видела во время обучения. Он проверяет способность модели обобщать и выдавать соответствующие результаты, не полагаясь на предыдущие примеры.
*  **Few-shot prompting** Здесь модели дается несколько примеров (shots), которые помогут ей отреагировать. Предоставляя контекст или предыдущие экземпляры, модель может лучше понять и сгенерировать желаемый результат. Например, можно показать модели несколько примеров переведенных предложений, прежде чем попросить ее перевести новое.
* **Chain-of-Thought (Цепочка мыслей)**. Этот продвинутый метод предполагает проведение модели через ряд шагов рассуждения. Разбивая сложную задачу на промежуточные этапы или «цепочки рассуждений», модель может добиться лучшего понимания языка и более точных результатов. Это похоже на пошаговое руководство для решения сложной математической задачи.

<div class="alert alert-info">

Быстрый прогресс в области обработки естественного языка (NLP) и широкое распространение больших языковых моделей (LLM) создали новую нишу и большой спрос на экспертов, которые могут создавать эффективные промпты. Эти люди, известные как **промпт инженеры**, являются не просто техническими специалистами, но и творцами, которые понимают нюансы языка, контекста и поведения ИИ.

Как сообщается в журнале Time, компании, от технологических гигантов до стартапов, осознают ценность специализированных должностей в области Prompt Engeneering'а. Поскольку решения на основе искусственного интеллекта все больше интегрируются в продукты и услуги, опыт **промпт инженера** гарантирует, что эти решения будут эффективными, удобными для пользователя и релевантными контексту.Такие сайты вакансий, как Indeed и LinkedIn, уже публикуют только в США тысячи вакансий на роль **промпт инженера** , с зарплатами от 50 000 до более 150 000 долларов в год. <br>
Российский рынок тоже начинает реагировать на новый мировой тренд, можете почитать [обзор](https://vc.ru/chatgpt/823927-kto-takie-prompt-inzhenery-i-mozhno-li-osvoit-etu-specialnost-ne-znaya-yazykov-programmirovaniya), там их называют **"промптёры"**.

# <center> 🧸 Выводы и заключения

Подводя итог, можно выделить несколько советов для базового промптинга, следуя которым можно сразу значительно поднять релевантность получаемых результатов.

1. Создайте адекватный промпт.
* **Ясность является ключевым моментом.** Убедитесь, что промпт ясен и недвусмысленнен. Избегайте жаргона, если это не необходимо для контекста.
* **Попробуйте сыграть в ролевую игру.** Как обсуждалось ранее, если модель возьмет на себя конкретную роль, это может привести к более адаптированным ответам.
* **Используйте ограничения.** Установка границ или ограничений может помочь направить модель к желаемому результату. Например, «Опишите Эйфелеву башню в трех предложениях» обеспечивает четкое ограничение длины.
* **Инструкция**. Вы можете разрабатывать эффективные промпты для различных простых задач, используя команды для указания модели, что вы хотите достичь, такие как "Write", "Classify", "Summarize", "Translate", "Order", и т.д. Еще одна рекомендация состоит в использовании некоторого ясного разделителя, например "###", для отделения инструкции и контекста.
* **Избегайте наводящих вопросов.** Наводящие вопросы могут исказить результаты модели. Очень важно сохранять нейтралитет, чтобы получить объективный ответ.
* **Избегайте неточностей**. Учитывая вышеуказанные рекомендации о детализации и улучшении формата, легко попасть в ловушку желания быть слишком умным в промптах и, возможно, создавать неточные описания. Часто лучше быть конкретным и прямым. Аналогия здесь очень похожа на эффективную коммуникацию - чем прямее, тем эффективнее передается сообщение. Возможно, с помощью таких промптов вы все равно получите неплохие ответы, но лучший промпт будет очень конкретным, кратким и по существу.
* **Делать или не делать?** - Избегайте формулировки того, что не нужно делать, а вместо этого указывать, что нужно делать. Это способствует большей специфичности и фокусу на деталях, которые приводят к хорошим результатам модели.

2. Повторяем и оцениваем

Процесс уточнения подсказок является итеративным. Вот типичный рабочий процесс:

* **Составьте начальный промпт.** В зависимости от поставленной задачи и желаемого результата.
* **Проверьте промпт.** Используйте модель ИИ для генерации ответа.
* **Оцените результат.** Проверьте, соответствует ли ответ намерению и критериям.
* **Уточните промпт.** Внесите необходимые корректировки на основе оценки.
* **Повторить.** Продолжайте этот процесс до тех пор, пока не будет достигнуто желаемое качество вывода.

3. Калибровка и точная настройка модели.

Помимо доработки самого промпта, существует также возможность калибровки или точной настройки модели ИИ. Это предполагает настройку параметров модели для лучшего соответствия конкретным задачам, наборам данных или экономике проекта (например, `temperature`, `top_n`, `max_tokens`). Хотя это более продвинутый метод, он может значительно улучшить производительность модели для специализированных приложений.


# 2_1_8


In [ ]:
import pandas as pd
path = 'https://stepik.org/media/attachments/lesson/1084297/submission100lines.csv'
df = pd.read_csv(path)

In [ ]:
df.iloc[0]

amount                                                     2
text_id                                             14205200
text       Сниму жилье. 500-600 рублей сутки. Эконом клас...
Name: 0, dtype: object

In [ ]:
contexts = [
    (
        'Сниму жилье. 500-600 рублей сутки. Эконом класса. Заезд 18.06. На 9-10 дней. На одного человека.'
    ),
    (
        "ищем жилье в центре недалеко от моря с 23.07-03.08 - нужен 1 двухместный номер, 1 трехместный, недорого. или как вариант дом на 2 семьи (5 чел)"
    ),
    (
        "Ищем жилье на период с 18-28июля..на две семьи 4взрослых 3 детей. .1500-2000в сутки за номер"
    )
]

In [ ]:
context_str

'Сниму жилье. 500-600 рублей сутки. Эконом класса. Заезд 18.06. На 9-10 дней. На одного человека.\n\n##\n\nищем жилье в центре недалеко от моря с 23.07-03.08 - нужен 1 двухместный номер, 1 трехместный, недорого. или как вариант дом на 2 семьи (5 чел)\n\n##\n\nИщем жилье на период с 18-28июля..на две семьи 4взрослых 3 детей. .1500-2000в сутки за номер'

In [ ]:
df.iloc[2]['text']

'Ищем жилье на период с 18-28июля..на две семьи 4взрослых 3 детей. .1500-2000в сутки за номер'

In [ ]:
contexts = df['text'][0:10]
context_str = '\n\n##\n\n'.join(contexts)
context_str

'Сниму жилье. 500-600 рублей сутки. Эконом класса. Заезд 18.06. На 9-10 дней. На одного человека.\n\n##\n\nищем жилье в центре недалеко от моря с 23.07-03.08 - нужен 1 двухместный номер, 1 трехместный, недорого. или как вариант дом на 2 семьи (5 чел)\n\n##\n\nИщем жилье на период с 18-28июля..на две семьи 4взрослых 3 детей. .1500-2000в сутки за номер\n\n##\n\n2 семьи по 3 человека (2 взрослых и ребенок) с 01.09-13.09 снимем жильё в пределах 5 минут от пляжа, 1500 руб/сутки. Рассмотрим все варианты.\n\n##\n\nСниму недорогое жилье в лазаревском на 6 чел 3 взр и 3 дет 30.07-09.08\n\n##\n\nИщем жилье на двоих с 7.08-14.08 . Ближе к центру и к моречку. (Wi-fi, кондиционер)\n\n##\n\nСниму жильё в Аше ( на улице Репина)\nС 16/29 июля \n2х местный писать в лс)\n\n##\n\nСнимем жилье с 30.06-7.07 Два взрослых и 2детей(7 и 11лет). Недалеко от моря.\n\n##\n\nСнимем жильё с 30 июня по 7 июля 3 взрослых и ребёнок\n\n##\n\nСниму не дорогое жилье 1 взрослый и 2 ребенка ( 9 лет и 5 лет). Примерно с 10 

In [ ]:
# Воспользуемся методом join для объединения контекстов в 1 строку с разделителем
# context_str = '\n\n##\n\n'.join(contexts)

contexts = df['text'][80:100]
context_str = '\n\n##\n\n'.join(contexts)

# Напишем промпт
prompt = f"""Ответь на вопрос, опираясь на контекст ниже.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь 'Я не знаю'.

###

Contexts:
{context_str}

###

Question: Даны строки с описанием заявок на жильё. Всего дано 20 строк.
Необходимо из текста заявки на жильё узнать информацию о количестве человек.
Нужно сложить всех людей в одной заявке.
Необходимо вернуть строку со значениями количеством человек для каждой заявки
через запятую. Без переноса строки. Т.е. у тебя должно быть 20 чисел через пробел.
Answer: """

llm = ChatOpenAI(temperature=0.0, course_api_key=course_api_key)
print(llm.predict(prompt))

2, 4, 4, 1, 2, 7, 3, 3, 3, 4, 5, 9, 9, 1, 2, 1, 2, 3, 1, 3


In [ ]:
x = [1, 5, 7, 6, 6, 2, 4, 5, 3, 5, 2, 6, 2, 4, 2, 6, 2, 4, 8, 6,
2, 4, 2, 3, 2, 2, 5, 2, 6, 2, 4, 8, 1, 3, 3, 3, 2, 2, 3, 3,
4, 4, 3, 3, 6, 2, 3, 2, 3, 2, 4, 4, 2, 2, 2, 2, 2, 2, 2, 2,
3, 4, 11, 3, 5, 2, 5, 3, 3, 3, 2, 2, 4, 5, 3, 8, 3, 2, 6, 1,
2, 4, 4, 1, 2, 7, 3, 3, 3, 4, 5, 9, 9, 1, 2, 1, 2, 3, 1, 3]

In [ ]:
df.head()

,amount,text_id,text
0,2,14205200,Сниму жилье. 500-600 рублей сутки. Эконом клас...
1,6,319097075,ищем жилье в центре недалеко от моря с 23.07-0...
2,4,98881311,Ищем жилье на период с 18-28июля..на две семьи...
3,4,44587027,2 семьи по 3 человека (2 взрослых и ребенок) с...
4,3,352802829,Сниму недорогое жилье в лазаревском на 6 чел 3...


In [ ]:
df['amount'] = x

In [ ]:
df.head()

,amount,text_id,text
0,1,14205200,Сниму жилье. 500-600 рублей сутки. Эконом клас...
1,5,319097075,ищем жилье в центре недалеко от моря с 23.07-0...
2,7,98881311,Ищем жилье на период с 18-28июля..на две семьи...
3,6,44587027,2 семьи по 3 человека (2 взрослых и ребенок) с...
4,6,352802829,Сниму недорогое жилье в лазаревском на 6 чел 3...


In [ ]:
df.to_csv('first_try.csv', index=False)

In [ ]:
template = """Ответь на вопрос, опираясь только на контекст ниже.
В ответе отправьте только число, без слов.
Если на вопрос нельзя ответить, используя информацию из контекста,
ответь '0'.

Context: {text_input}

Question: Для какого количества человек требуется жилье (осторожно, не учитывай одного человека несколько раз, вчитайся в текст)?

Answer:
"""

prompt = PromptTemplate(template=template, input_variables=["text_input"])
openai_llm = ChatOpenAI(temperature=0.0, max_tokens=16, course_api_key=course_api_key)
llm_chain = LLMChain(prompt=prompt, llm=openai_llm)

# amount_list = []
# for text_input in df_sub['text']:
#     amount_list.append(llm_chain.run(text_input))

# df_sub['amount'] = np.array(list(map(int, amount_list)))
# df_sub.to_csv('submission_2.1.8.csv', index=False)